apps/mobile/src/hooks/useMentionableMessage.ts

```diff
@@ -0,0 +1,178 @@
+import useDebounce from '~/shared/hooks/useDebounce';
+
+type MentionDataType = {
+  interval: {
+    start: number;
+    length: number;
+  };
+  userId?: string;
+  communityId?: string;
+};
+
+export type MentionType = {
+  type: 'User' | 'Community';
+  label: string;
+  value: string;
+};
+
+type Mention = {
+  interval: {
+    start: number;
+    length: number;
+  };
+  userId?: string;
+  communityId?: string;
+};
+
+export function useMentionableMessage(queryRef: useMentionableMessageQueryFragment$key) {
+  const query = useFragment(
+    graphql`
+      fragment useMentionableMessageQueryFragment on Query {
+        ...isFeatureEnabledFragment
+      }
+    `,
+    queryRef
+  );
+
+  const [message, setMessage] = useState('');
+  const [mentions, setMentions] = useState<MentionDataType[]>([]);
+
+  const [isSelectingMentions, setIsSelectingMentions] = useState(false);
+
+  const [aliasKeyword, setAliasKeyword] = useState('');
+  const debouncedAliasKeyword = useDebounce(aliasKeyword, 100);
+
+  const [selection, setSelection] = useState({ start: 0, end: 0 });
+
+  const isMentionEnabled = isFeatureEnabled(FeatureFlag.MENTIONS, query);
+
+  const handleSetMention = useCallback(
+    (mention: MentionType) => {
+      // Use substring to only look up to the current cursor position
+      const upToCursor = message.substring(0, selection.start);
+      const mentionStartPos = upToCursor.lastIndexOf(aliasKeyword);
+      const mentionEndPos = mentionStartPos + aliasKeyword.length;
+
+      const newMessage = `${message.substring(0, mentionStartPos)}@${
+        mention.label
+      } ${message.substring(mentionEndPos)}`;
+
+      setMessage(newMessage);
+
+      const newMention: Mention = {
+        interval: {
+          start: mentionStartPos,
+          length: mention.label.length + 1, // +1 for the @
+        },
+      };
+
+      if (mention.type === 'User') {
+        newMention.userId = mention.value;
+      } else {
+        newMention.communityId = mention.value;
+      }
+
+      // Calculate the length difference between the old alias and the new mention
+      const lengthDifference = mention.label.length + 2 - aliasKeyword.length; // +1 for the @
+
+      // Adjust the positions of mentions that come after the newly added mention
+      const adjustedMentions = mentions.map((existingMention) => {
+        if (existingMention.interval.start >= mentionStartPos) {
+          return {
+            ...existingMention,
+            interval: {
+              start: existingMention.interval.start + lengthDifference,
+              length: existingMention.interval.length,
+            },
+          };
+        }
+        return existingMention;
+      });
+
+      setIsSelectingMentions(false);
+
+      setMentions([...adjustedMentions, newMention]);
+    },
+    [mentions, message, setMessage, aliasKeyword, selection.start]
+  );
+
+  const handleSetMessage = useCallback(
```

Comment:
imo i think the logic in `handleSetMention` and `handleSetMessage` should be unit tested

Reply:
Agree! Will write the unit test for this hooks